pkg/apis/networking/validation/validation.go

```diff
@@ -277,26 +312,24 @@ func validateHTTPIngressRuleValue(httpIngressRuleValue *networking.HTTPIngressRu
 	if len(httpIngressRuleValue.Paths) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("paths"), ""))
 	}
-	for i, rule := range httpIngressRuleValue.Paths {
-		if len(rule.Path) > 0 {
-			if !strings.HasPrefix(rule.Path, "/") {
-				allErrs = append(allErrs, field.Invalid(fldPath.Child("paths").Index(i).Child("path"), rule.Path, "must be an absolute path"))
-			}
-			// TODO: More draconian path regex validation.
-			// Path must be a valid regex. This is the basic requirement.
-			// In addition to this any characters not allowed in a path per
-			// RFC 3986 section-3.3 cannot appear as a literal in the regex.
-			// Consider the example: http://host/valid?#bar, everything after
-			// the last '/' is a valid regex that matches valid#bar, which
-			// isn't a valid path, because the path terminates at the first ?
-			// or #. A more sophisticated form of validation would detect that
-			// the user is confusing url regexes with path regexes.
-			_, err := regexp.CompilePOSIX(rule.Path)
-			if err != nil {
-				allErrs = append(allErrs, field.Invalid(fldPath.Child("paths").Index(i).Child("path"), rule.Path, "must be a valid regex"))
+	for i, path := range httpIngressRuleValue.Paths {
+		idxPath := fldPath.Child("paths").Index(i)
+		if !strings.HasPrefix(path.Path, "/") {
+			allErrs = append(allErrs, field.Invalid(idxPath.Child("path"), path.Path, "must be an absolute path"))
+		}
+
+		if path.PathType != nil {
+			if !supportedPathTypes.Has(string(*path.PathType)) {
+				allErrs = append(allErrs, field.NotSupported(idxPath.Child("pathType"), path.PathType, supportedPathTypes.List()))
 			}
+
+			if *path.PathType == networking.PathTypeExact || *path.PathType == networking.PathTypePrefix {
+				if len(path.Path) > 0 && strings.Contains(path.Path, "//") {
```

Comment:
let's factor this out

```
validateIngressPath(path string)  error

switch *path.PathType {
  case networking.PathTypeExact:
    allErrs = append(allErrs, validateIngressPath(...))
  case networking.PathTypePrefix:
    allErrs = append(validateIngressPath(...))
  case  networking.ImplementationSpecific:
   ...
  default:
   // error
}

Reply:
I reworked this to use a switch statement. There was only one real condition for it, but I think it's still an improvement over what I had initially. Thanks!